src/WindowsUtils/Debugger.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "WindowsUtils/Debugger.h"
+
+#include "OrbitBase/GetLastError.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/StringConversion.h"
+#include "OrbitBase/ThreadUtils.h"
+#include "WindowsUtils/CreateProcess.h"
+#include "WindowsUtils/SafeHandle.h"
+
+namespace orbit_windows_utils {
+
+namespace {
+
+ErrorMessageOr<Debugger::StartInfo> CreateStartInfoOrError(
+    const ErrorMessageOr<ProcessInfo>& process_info_or_error) {
+  if (process_info_or_error.has_error()) return process_info_or_error.error();
+  const ProcessInfo& process_info = process_info_or_error.value();
+  Debugger::StartInfo start_info;
+  start_info.working_directory = process_info.working_directory;
+  start_info.command_line = process_info.command_line;
+  start_info.process_id = process_info.process_id;
+  return start_info;
+}
+
+}  // namespace
+
+Debugger::Debugger(std::vector<DebugEventListener*> debug_event_listeners)
+    : debug_event_listeners_(std::move(debug_event_listeners)) {}
+
+Debugger::~Debugger() { Wait(); }
+
+ErrorMessageOr<Debugger::StartInfo> Debugger::Start(const std::filesystem::path& executable,
+                                                    const std::filesystem::path& working_directory,
+                                                    const std::string_view arguments) {
+  // Make sure we have at least on debug event listener.
+  if (debug_event_listeners_.size() == 0) {
+    return ErrorMessage("Debugger has no debug event listener");
+  }
+
+  // Launch process and start debugging loop on the same separate thread.
+  thread_ = std::thread(&Debugger::DebuggerThread, this, executable, working_directory,
+                        std::string(arguments));
+
+  // Return "StartInfo" or error based on the result of process creation from "DebuggerThread".
+  return start_info_or_error_promise_.GetFuture().Get();
```

Comment:
Nit: So this blocks, right? I would add that to the comment.

Reply:
Modified the comment.